Race condition triggered in java-generator for very big schemas

### Describe the bug

Using different versions of Java produce different errors with the `java-generator`

### Fabric8 Kubernetes Client version

SNAPSHOT

### Steps to reproduce

- `curl -Ls https://raw.githubusercontent.com/istio/istio/master/manifests/charts/base/crds/crd-all.gen.yaml -o istio.yaml`
- `sdk use java 11.0.17-tem`
- `jbang io.fabric8:java-generator-cli:6.3.1 --add-extra-annotations --source istio.yaml --target ./src`

Produces a stack-trace like:
```
Exception in thread "main" java.lang.AssertionError
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
        at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:603)
        at java.base/java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:678)
        at java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:737)
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:159)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:173)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
        at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
        at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:661)
        at io.fabric8.java.generator.CRGeneratorRunner.runOnSingleSource(CRGeneratorRunner.java:82)
        at io.fabric8.java.generator.CRGeneratorRunner.run(CRGeneratorRunner.java:66)
        at io.fabric8.java.generator.cli.GenerateJavaSources.run(GenerateJavaSources.java:77)
        at picocli.CommandLine.executeUserObject(CommandLine.java:2026)
        at picocli.CommandLine.executeHelpRequest(CommandLine.java:2012)
        at picocli.CommandLine.executeHelpRequest(CommandLine.java:1983)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2268)
        at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
        at picocli.CommandLine.execute(CommandLine.java:2170)
        at io.fabric8.java.generator.cli.GenerateJavaSources.main(GenerateJavaSources.java:81)
Caused by: java.lang.AssertionError
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
        at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:603)
        at java.base/java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:678)
        at java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:737)
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:159)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:173)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
        at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
        at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:661)
        at io.fabric8.java.generator.CRGeneratorRunner.lambda$runOnSingleSource$2(CRGeneratorRunner.java:93)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
        at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:290)
        at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:746)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.lang.AssertionError: I am not a child of my parent.
        at com.github.javaparser.printer.DefaultPrettyPrinterVisitor.printOrphanCommentsBeforeThisChildNode(DefaultPrettyPrinterVisitor.java:1799)
        at com.github.javaparser.printer.DefaultPrettyPrinterVisitor.visit(DefaultPrettyPrinterVisitor.java:1584)
        at com.github.javaparser.printer.DefaultPrettyPrinterVisitor.visit(DefaultPrettyPrinterVisitor.java:53)
        at com.github.javaparser.ast.expr.NormalAnnotationExpr.accept(NormalAnnotationExpr.java:74)
        at com.github.javaparser.printer.DefaultPrettyPrinterVisitor.printMemberAnnotations(DefaultPrettyPrinterVisitor.java:94)
        at com.github.javaparser.printer.DefaultPrettyPrinterVisitor.visit(DefaultPrettyPrinterVisitor.java:268)
        at com.github.javaparser.printer.DefaultPrettyPrinterVisitor.visit(DefaultPrettyPrinterVisitor.java:53)
        at com.github.javaparser.ast.body.ClassOrInterfaceDeclaration.accept(ClassOrInterfaceDeclaration.java:98)
        at com.github.javaparser.printer.DefaultPrettyPrinterVisitor.visit(DefaultPrettyPrinterVisitor.java:215)
        at com.github.javaparser.printer.DefaultPrettyPrinterVisitor.visit(DefaultPrettyPrinterVisitor.java:53)
        at com.github.javaparser.ast.CompilationUnit.accept(CompilationUnit.java:133)
        at com.github.javaparser.printer.DefaultPrettyPrinter.print(DefaultPrettyPrinter.java:95)
        at com.github.javaparser.ast.Node.toString(Node.java:322)
        at io.fabric8.java.generator.WritableCRCompilationUnit.writeAllJavaClasses(WritableCRCompilationUnit.java:58)
        at io.fabric8.java.generator.CRGeneratorRunner.lambda$runOnSingleSource$1(CRGeneratorRunner.java:93)
        ... 10 more
```

Additional issues can be experienced by mixing and matching Java and `kubernetes-client` versions, another example stack trace is:
```
java.lang.ArrayIndexOutOfBoundsException
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
        at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:603)
        at java.base/java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:678)
        at java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:737)
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:159)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:173)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
        at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
        at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:661)
        at io.fabric8.java.generator.FileJavaGenerator.runOnSingleSource(FileJavaGenerator.java:87)
        at io.fabric8.java.generator.FileJavaGenerator.run(FileJavaGenerator.java:71)
        at io.fabric8.java.generator.cli.GenerateJavaSources.lambda$run$0(GenerateJavaSources.java:132)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at io.fabric8.java.generator.cli.GenerateJavaSources.run(GenerateJavaSources.java:132)
        at picocli.CommandLine.executeUserObject(CommandLine.java:2026)
        at picocli.CommandLine.executeHelpRequest(CommandLine.java:2012)
        at picocli.CommandLine.executeHelpRequest(CommandLine.java:1983)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2268)
        at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
        at picocli.CommandLine.execute(CommandLine.java:2170)
        at io.fabric8.java.generator.cli.GenerateJavaSources.main(GenerateJavaSources.java:136)
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 8 out of bounds for length 7
        at java.base/java.util.ArrayList.add(ArrayList.java:487)
        at java.base/java.util.ArrayList.add(ArrayList.java:499)
        at com.github.javaparser.ast.Node.setParentNode(Node.java:435)
        at com.github.javaparser.ast.NodeList.setAsParentNodeOf(NodeList.java:559)
        at com.github.javaparser.ast.NodeList.own(NodeList.java:81)
        at com.github.javaparser.ast.NodeList.add(NodeList.java:73)
        at com.github.javaparser.ast.nodeTypes.NodeWithAnnotations.addAnnotation(NodeWithAnnotations.java:59)
        at io.fabric8.java.generator.nodes.JObjectExtraAnnotations.addExtraAnnotations(JObjectExtraAnnotations.java:36)
        at io.fabric8.java.generator.nodes.JObject.generateJava(JObject.java:156)
        at io.fabric8.java.generator.nodes.JObject.generateJava(JObject.java:168)
        at io.fabric8.java.generator.nodes.JArray.generateJava(JArray.java:51)
        at io.fabric8.java.generator.nodes.JObject.generateJava(JObject.java:168)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
        at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
        at io.fabric8.java.generator.CRGeneratorRunner.validateAndAggregate(CRGeneratorRunner.java:112)
        at io.fabric8.java.generator.CRGeneratorRunner.generate(CRGeneratorRunner.java:97)
        at io.fabric8.java.generator.FileJavaGenerator.lambda$runOnSingleSource$3(FileJavaGenerator.java:95)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
        at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:290)
        at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:746)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```

### Expected behavior

The generator correctly run and generate the sources.

### Runtime

other (please specify in additional context)

### Kubernetes API Server version

1.25.3@latest

### Environment

Windows, Linux, macOS

### Fabric8 Kubernetes Client Logs

_No response_

### Additional context

_No response_

On which version of java are you getting this stacktrace?

> Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 8 out of bounds for length 7
        at java.base/java.util.ArrayList.add(ArrayList.java:487)
        at java.base/java.util.ArrayList.add(ArrayList.java:499)
        at com.github.javaparser.ast.Node.setParentNode(Node.java:435)

Ideally you should be able to isolate a simplified test case that reproduces the issue because otherwise it will be difficult to help you.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,7 @@
 * Fix #4798: fix leader election release on cancel
 * Fix #4815: (java-generator) create target download directory if it doesn't exist
 * Fix #4818: [java-generator] Escape `*/` in generated JavaDocs
+* Fix #4723: [java-generator] Fix a race in the use of JavaParser hitting large CRDs
 
 #### Improvements
 * Fix #4747: migrate to SnakeYAML Engine

--- a/java-generator/cli/src/main/java/io/fabric8/java/generator/cli/GenerateJavaSources.java
+++ b/java-generator/cli/src/main/java/io/fabric8/java/generator/cli/GenerateJavaSources.java
@@ -75,7 +75,7 @@ public class GenerateJavaSources implements Runnable {
   String codeStructure = null;
 
   @Option(names = { "-skip-generated-annotations",
-      "--skip-generated-annotations" }, description = "Add extra lombok and sundrio annotation to the generated classes", required = false, hidden = true)
+      "--skip-generated-annotations" }, description = "Skip emitting the @javax.annotation.processing.Generated annotation on the generated sources", required = false, hidden = true)
   Boolean skipGeneratedAnnotations = null;
 
   @Option(names = { "-package-overrides",
@@ -87,7 +87,7 @@ public class GenerateJavaSources implements Runnable {
     final Config.Prefix pSt = (prefixStrategy != null) ? Config.Prefix.valueOf(prefixStrategy) : null;
     final Config.Suffix sSt = (suffixStrategy != null) ? Config.Suffix.valueOf(suffixStrategy) : null;
     final Config.CodeStructure structure = (codeStructure != null) ? Config.CodeStructure.valueOf(codeStructure) : null;
-    final Boolean generatedAnnotations = (skipGeneratedAnnotations != null) ? skipGeneratedAnnotations : null;
+    final Boolean noGeneratedAnnotations = (skipGeneratedAnnotations != null) ? skipGeneratedAnnotations : false;
     final Config config = new Config(
         uppercaseEnum,
         pSt,
@@ -95,7 +95,7 @@ public class GenerateJavaSources implements Runnable {
         alwaysPreserveUnkownFields,
         addExtraAnnotations,
         structure,
-        generatedAnnotations,
+        !noGeneratedAnnotations,
         packageOverrides);
 
     List<JavaGenerator> runners = new ArrayList<>();

--- a/java-generator/core/pom.xml
+++ b/java-generator/core/pom.xml
@@ -49,13 +49,6 @@
 
     <dependency>
       <groupId>io.sundr</groupId>
-      <artifactId>sundr-adapter-reflect</artifactId>
-      <version>${sundrio.version}</version>
-      <scope>compile</scope>
-    </dependency>
-
-    <dependency>
-      <groupId>io.sundr</groupId>
       <artifactId>builder-annotations</artifactId>
       <version>${sundrio.version}</version>
       <scope>compile</scope>

--- a/java-generator/core/src/main/java/io/fabric8/java/generator/CRGeneratorRunner.java
+++ b/java-generator/core/src/main/java/io/fabric8/java/generator/CRGeneratorRunner.java
@@ -98,7 +98,7 @@ public class CRGeneratorRunner {
 
       List<GeneratorResult.ClassResult> classResults = validateAndAggregate(crGenerator, specGenerator, statusGenerator);
 
-      writableCUs.add(new WritableCRCompilationUnit(classResults));
+      writableCUs.add(new WritableCRCompilationUnit(classResults, basePackageName));
     }
 
     return writableCUs;

--- a/java-generator/core/src/main/java/io/fabric8/java/generator/FileJavaGenerator.java
+++ b/java-generator/core/src/main/java/io/fabric8/java/generator/FileJavaGenerator.java
@@ -95,10 +95,9 @@ public class FileJavaGenerator implements JavaGenerator {
                     CustomResourceDefinition crd = (CustomResourceDefinition) resource;
 
                     final String basePackage = groupToPackage(crd.getSpec().getGroup());
-                    List<WritableCRCompilationUnit> writables = crGeneratorRunner.generate(crd, basePackage);
 
-                    writables.parallelStream()
-                        .forEach(w -> w.writeAllJavaClasses(basePath, basePackage));
+                    crGeneratorRunner.generate(crd, basePackage).parallelStream()
+                        .forEach(w -> w.writeAllJavaClasses(basePath));
                   } else {
                     LOGGER.warn("Not generating nothing for resource of kind: {}", resource.getKind());
                   }

--- a/java-generator/core/src/main/java/io/fabric8/java/generator/WritableCRCompilationUnit.java
+++ b/java-generator/core/src/main/java/io/fabric8/java/generator/WritableCRCompilationUnit.java
@@ -34,16 +34,18 @@ public class WritableCRCompilationUnit {
   private static final Logger LOGGER = LoggerFactory.getLogger(WritableCRCompilationUnit.class);
 
   private final List<GeneratorResult.ClassResult> classResults;
+  private final String basePackage;
 
-  WritableCRCompilationUnit(List<GeneratorResult.ClassResult> classResults) {
+  WritableCRCompilationUnit(List<GeneratorResult.ClassResult> classResults, String basePackage) {
     this.classResults = classResults;
+    this.basePackage = basePackage;
   }
 
   public List<GeneratorResult.ClassResult> getClassResults() {
     return classResults;
   }
 
-  public void writeAllJavaClasses(File basePath, String basePackage) {
+  public void writeAllJavaClasses(File basePath) {
     try {
       createFolders(basePackage, basePath);
       for (GeneratorResult.ClassResult cr : this.classResults) {

--- a/java-generator/core/src/main/java/io/fabric8/java/generator/nodes/AbstractJSONSchema2Pojo.java
+++ b/java-generator/core/src/main/java/io/fabric8/java/generator/nodes/AbstractJSONSchema2Pojo.java
@@ -42,9 +42,11 @@ public abstract class AbstractJSONSchema2Pojo {
   static final String OBJECT_CRD_TYPE = "object";
   static final String ARRAY_CRD_TYPE = "array";
 
-  public static final AnnotationExpr GENERATED_ANNOTATION = new SingleMemberAnnotationExpr(
-      new Name("javax.annotation.processing.Generated"),
-      new StringLiteralExpr("io.fabric8.java.generator.CRGeneratorRunner"));
+  public static final AnnotationExpr newGeneratedAnnotation() {
+    return new SingleMemberAnnotationExpr(
+        new Name("javax.annotation.processing.Generated"),
+        new StringLiteralExpr("io.fabric8.java.generator.CRGeneratorRunner"));
+  }
 
   protected final String description;
   protected final Config config;

--- a/java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JCRObject.java
+++ b/java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JCRObject.java
@@ -140,7 +140,7 @@ public class JCRObject extends AbstractJSONSchema2Pojo implements JObjectExtraAn
     }
 
     if (config.isGeneratedAnnotations()) {
-      clz.addAnnotation(GENERATED_ANNOTATION);
+      clz.addAnnotation(newGeneratedAnnotation());
     }
     if (config.isObjectExtraAnnotations()) {
       addExtraAnnotations(clz);

--- a/java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JObject.java
+++ b/java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JObject.java
@@ -150,7 +150,7 @@ public class JObject extends AbstractJSONSchema2Pojo implements JObjectExtraAnno
                 "using = com.fasterxml.jackson.databind.JsonDeserializer.None.class")));
 
     if (config.isGeneratedAnnotations()) {
-      clz.addAnnotation(GENERATED_ANNOTATION);
+      clz.addAnnotation(newGeneratedAnnotation());
     }
     if (config.isObjectExtraAnnotations()) {
       addExtraAnnotations(clz);

--- a/java-generator/core/src/test/java/io/fabric8/java/generator/GeneratorTest.java
+++ b/java-generator/core/src/test/java/io/fabric8/java/generator/GeneratorTest.java
@@ -452,6 +452,7 @@ class GeneratorTest {
         null,
         Boolean.FALSE,
         null);
+    String generatedAnnotationName = AbstractJSONSchema2Pojo.newGeneratedAnnotation().getNameAsString();
 
     // Act
     GeneratorResult res1 = obj1.generateJava();
@@ -459,10 +460,10 @@ class GeneratorTest {
 
     // Assert
     Optional<ClassOrInterfaceDeclaration> clz1 = res1.getTopLevelClasses().get(0).getCompilationUnit().getClassByName("T");
-    assertTrue(clz1.get().getAnnotationByName(AbstractJSONSchema2Pojo.GENERATED_ANNOTATION.getNameAsString()).isPresent());
+    assertTrue(clz1.get().getAnnotationByName(generatedAnnotationName).isPresent());
 
     Optional<ClassOrInterfaceDeclaration> clz2 = res2.getTopLevelClasses().get(0).getCompilationUnit().getClassByName("T");
-    assertFalse(clz2.get().getAnnotationByName(AbstractJSONSchema2Pojo.GENERATED_ANNOTATION.getNameAsString()).isPresent());
+    assertFalse(clz2.get().getAnnotationByName(generatedAnnotationName).isPresent());
   }
 
   @Test

